### DISCUSSION TITLE
Hosting a plausible instance with nginx

### DISCUSSION THREAD

**Discussion by lmrodriguezr** [plausible/analytics#4189] [Self-Hosted Support]
Dear Plausible team,

Thank you for the amazing product! `v2` has simplified a lot!

I'm currently trying to setup an instance of Plausible in a machine that is already serving web using NGINX, and I've not been able to configure it properly. The current server already uses HTTPS and it's working properly, but the plausible instance doesn't seem to see the outside world. The changes I've made are:

`plausible-conf.env`
```diff
@@ -1,3 +1,3 @@
-BASE_URL=replace-me
-SECRET_KEY_BASE=replace-me
-TOTP_VAULT_KEY=replace-me
+BASE_URL=https://my-actual-url.com
+SECRET_KEY_BASE=MY/ACTUAL/SECRET
+TOTP_VAULT_KEY=MY/ACTUAL/SECRET
```

Then I followed the instructions in https://github.com/plausible/community-edition/tree/v2.1.0/reverse-proxy#nginx (except for the `certbot` command, because I have not installed the `nginx` plugin, but it's already configured to cover HTTPS for the domain name I'm using). Then I simply ran:

`sudo docker compose -f docker-compose.yml up -d`

Which seems to work properly (Running, Running, and Started). And port 8000 seems to be listening:

```
$ sudo netstat -tulpn | grep 8000
tcp        0      0 127.0.0.1:8000          0.0.0.0:*               LISTEN      2728818/docker-prox 
```

But when I try to access the site I cannot get to it:

```
$ curl https://localhost:8000
curl: (35) OpenSSL SSL_connect: SSL_ERROR_SYSCALL in connection to localhost:8000
```

And if I try using the actual domain name (instead of `localhost`) I get:
```
curl: (7) Failed to connect to ..... port 8000: Connection refused
```

It's as if it's not listening at all from that port, but I'm not sure what I'm doing wrong.

Any help is greatly appreciated!
Miguel.

**Comment by ruslandoga** [comment]
👋 @lmrodriguezr 

What about `curl http://localhost:8000` (http, not https)?

**Reply by ruslandoga** [reply]
And `docker compose logs`

**Reply by lmrodriguezr** [reply]
```
$ curl http://localhost:8000
curl: (56) Recv failure: Connection reset by peer
```

And the `compose logs`: https://gist.github.com/lmrodriguezr/4e1644a8c15c0752e45a8a838ab91051

Thanks!

**Reply by ruslandoga** [reply]
The problem is with PostgreSQL.

> FATAL:  database files are incompatible with server
> DETAIL:  The data directory was initialized by PostgreSQL version 14, which is not compatible with this version 16.3.

You updated PostgreSQL's major version without performing an [upgrade.](https://github.com/plausible/community-edition/blob/v2.1.0/upgrade/postgres.md) There are many alternative fixes, you can search the internet or ask a GPT for them :) The quickest is to go back to v14.

**Reply by lmrodriguezr** [reply]
Thank you @ruslandoga !
With your help, it seems I'm awfully close now, but not quite at the finish line. I was able to successfully launch the service, and I configured NGINX to serve it via HTTP and HTTPS in a subdomain. Indeed, the webpage loads nicely from my browser, but I cannot do anything in the portal. After inspecting the browser's console, I found out this error is popping up every second or so:

```
WebSocket connection to 'wss://analytics.my-domain.com/live/websocket?_csrf_token=...' failed:
  connect             @ app-...
  (anonymous)         @ app-...
  (anonymous)         @ app-...
  waitForSocketClosed @ app-...
  (anonymous)         @ app-...
  waitForBufferDone   @ app-... 
  teardown            @ app-...
  (anonymous)         @ app-...
  (anonymous)         @ app-...
...
```


This is my nginx `plausible.conf` file:

```
server {
  server_name analytics.my-domain.com;
  ssl_certificate ......./fullchain.pem;
  ssl_certificate_key .......privkey.pem;

  listen 80;
  listen [::]:80;
  listen 443 ssl;
  ssl on;

  location / {
    proxy_pass http://127.0.0.1:8000;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

    location = /live/websocket {
      proxy_pass http://127.0.0.1:8000;
      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "Upgrade";
    }
  }
}
```

Any ideas would be very welcomed! :)
Thanks!
Miguel.

**Reply by ruslandoga** [reply]
You can compare your nginx configuration to the example one in plausible/community-edition repo. You can also check for any errors in docker compose logs (it might be a base_url issue).

**Reply by lmrodriguezr** [reply]
Thank you, it works now! In case somebody else finds the same issue: The configuration was correct, but there was some hang-up on cache (not sure where, tho). After restarting nginx a couple more times, and emptying the local browser cache, it all worked out.

**Reply by samjohn77** [reply]
@lmrodriguezr @ruslandoga  I am facing the same issue and am not able to resolve it even after restarting Nginx multiple times. The URL wss://example-domain.com/live/websocket?_csrf_token=SJtbL2Agh&_mounts=0&_live_referer=undefined&vsn=2.0.0 is returning a 403 error.

**Reply by ruslandoga** [reply]
Please share your `docker compose logs`

**Reply by samjohn77** [reply]
@ruslandoga  
I am really sorry for wasting your time. I should have checked the Docker Compose logs. I figured out the error myself—the base URL was pointing to a different subdomain that was not configured. Thank you so much for your time and support. God bless you! If my project is successful, I plan to share my revenue with good OSS projects like this one. Right now, I am just starting my project, so I am relying on OSS self-hosting.
